docs/source/reference/types.rst

```diff
@@ -0,0 +1,286 @@
+To write this, we use the syntax
+``interface SPEC | PARAMS``, where ``PARAMS`` are determining.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Desguaring and Constructors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+An interface is desugared roughly as follows:
+
+1. Take the interface, turn it into a record, and make all its methods into arguments to the record
+2. Turn every instance of the interface into a definition of the record.
+3. For each named one, assign the name to the definition, for each unnamed one, add the ``%hint`` pragma
+4. Transform every case of ``f : i a => b`` to ``f : {_ : i' a} -> b``, and pattern match on the implicit argument in the definition.
+
+Given this, it seems natural to ask how we can construct "anonymous" inline instances.
+Idris takes this into account, by providing interface constructors.
+An interface constructor extends an interface by allowing a ``constructor NAME`` clause in an interface definition, which creates a "interface value" constructor. For instance, the following two implementations are equivalent (given `semigroup`_):
+
+.. code:: idris
+
+    [semiList] {a : Type} -> Semigroup (List a) where
+        (<+>) = (++)
+
+and
+
+.. code:: idris
+
+    semiList : {a : Type} -> Semigroup (List a)
+    semiList = (MkSemigroup (++))
+
+----------------------
+Other Forms
+----------------------
+
+The two main ways to declare types at the top level are ``record`` and ``data``
+
+.. note::
+    Type synonyms, type families, and the like from Haskell are all just normal functions from ``Type`` to something else.
+    In addition, this means that one can have anonymous instances, so, for instance, ``Monad (\x => x)`` is a perfectly valid form.
+
+    Even so, it is in general not a good idea to create such instances
+
+``data`` declares (General) Algebraic Data Types ((G)ADTs).
+The Haskell style, simple ADT form is always a statement of the form ``data NAME ARGS = ALT {| ALT}``.
+``ARGS`` must be a valid list of names.
+In addition, each ``ALT`` must be of the form ``CONS PARAMS`` where each ``CONS`` is either an operator or a name (that will be declared as a constructor).
+Finally, each of ``PARAMS`` must be a valid type.
+
+To declare (Agda style) Generalized Algebraic Data Types, we must write ``data NAME : KIND where CONS``.
+``KIND`` must be a kind (in Idris there isn't actually distinction between types and kinds) whose tail ultimately ends up returning the type ``Type``.
+So, for instance, ``data Vect : (n : Nat) -> Type where ...`` is valid, whilst ``data Vect : forall k. (n : Nat) -> k where ...`` is not.
+Each ``CONS`` must be a function declaration, except that the name being defined must start with an uppercase letter and the ultimate return type of it must be a fully applied instance of ``NAME``.
+
+The other way to define top level types, ``record``, is documented in detail on the `records`_ page.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Existentials in Types
```

Comment:
```suggestion
Inferable Placeholders
```

I don't know what to call this class of terms but `existential` is extremely confusing because there is already something called `existential types` in programming and it's not what is exposed here. Existential types are usually an encoding of sigma-types for languages without dependent types.